DRILL-6960: AutoLimit the size of ResultSet for a WebUI (or REST) client

kkhatua opened the pull request:
Fixes the bug introduced with DRILL-6050 (#1593 )
1. Check if query can be wrapped with a limit and provide a warning if option was selected
2. Switch `help` from onclick to a hovering tooltip

Screenshot:
![image](https://user-images.githubusercontent.com/4335237/51008596-4f3c1900-1501-11e9-93b9-c78a1b333ecb.png)

kkhatua commented:
@ihuzenko can you review this?

ihuzenko commented:
@kkhatua , thanks for making changes I'll take a look soon.

ihuzenko commented:
@kkhatua In my opinion it would be better to implement the feature on the back-end side. In such case UI only should validate that when 'Limit results to' checkbox is enabled user must provide valid integer number (only digits without any other signs). For the backend side I can't give you full solution now, but you could try something like this inside DrillSqlWorker.getQueryPlan: 

```
if (context.isAutoLimitEnabled() && sqlNode instanceof SqlSelect) {
      
  String limit = context.getAutoLimitRowCount(); // limit passed from UI
      
  SqlNumericLiteral fetch = SqlNumericLiteral.createExactNumeric(limit, sqlNode.getParserPosition());
      
  ((SqlSelect)sqlNode).setFetch(fetch);
}
```

kkhatua commented:
The original feature in #1593 was a workaround for the lack of pagination from the server side, when queries are submitted through the WebUI. The autoWrapping is actually not something we want on the server end. This is just a way to ensure that there is an option to not submit a query that can return a very large JSON response and potentially hang/crash the Drillbit.
Hence, the logic was done completely on the front-end and without any changes to the REST APIs.

arina-ielchiieva commented:
@kkhatua I also agree with Igor. Parsing query on the client side is error prone (I already can think of two cases that was missed: values queries and multiline comments) and looks like overhead when on server side we have all means to add query limit without line by line parsing. You don't need to change existing REST API, just add new call for Web UI that would pass limit along with the query. You can also take a look at JDBC calls where limit 0 is added for the queries to find out data types.

kkhatua commented:
@ihuzenko / @arina-ielchiieva / @vdiravka 
I've made changes within this PR to capture the requirements.
The automatic limitting of the resultset is now handled by the server, and involved making changes to the Protobuf as well.
Here are the changes done for the latest commit to this PR.
1. Submit autoLimit value to server
2. Server is able to detect the autoLimit and apply the wrapper around the input SQL.
3. If autoLimit is not applicable, remove the value to avoid incorrectly marking the profile during execution.
4. Optional AutoLimit value is set in the profile if autoLimit is being applied
5. WebUI will render a pinned warning indicating that the result set size was artificially limited (NOTE: this is a workaround since QueryManager cannot modify the actually query text with header comments, etc.)

kkhatua commented:
Screenshot for profile when query is submitted via WebUI with autoLimiting to 23 rows:
Pinned message at Query Text
![image](https://user-images.githubusercontent.com/4335237/51502352-9b514e00-1d8a-11e9-91a9-92021d30915f.png)

Pinned message at Operator Overview (since rows are shown here)
![image](https://user-images.githubusercontent.com/4335237/51502390-c2a81b00-1d8a-11e9-8dfd-ff913de1f883.png)

Screenshot of profile where non-select query was submitted with auto-limit of 3 (no pinned messages):
![image](https://user-images.githubusercontent.com/4335237/51502415-de132600-1d8a-11e9-8487-be7a3243549b.png)

Console output
```
2019-01-21 13:26:09,432 kk127.lab [23b9c70d-cb96-dd12-26b7-1380e3544bda:foreman] INFO  o.a.drill.exec.work.foreman.Foreman - Query text for query with id 23b9c70d-cb96-dd12-26b7-1380e3544bda issued by mapr: show databases
2019-01-21 13:26:09,436 kk127.lab [23b9c70d-cb96-dd12-26b7-1380e3544bda:foreman] DEBUG  o.a.d.e.planner.sql.DrillSqlWorker - Disabled Auto limit for show databases
2019-01-21 13:26:09,744 kk127.lab [23b9c70d-cb96-dd12-26b7-1380e3544bda:frag:0:0] INFO  o.a.d.e.w.fragment.FragmentExecutor - 23b9c70d-cb96-dd12-26b7-1380e3544bda:0:0: State change requested AWAITING_ALLOCATION --> RUNNING
2019-01-21 13:26:09,766 kk127.lab [23b9c70d-cb96-dd12-26b7-1380e3544bda:frag:0:0] INFO  o.a.d.e.w.fragment.FragmentExecutor - 23b9c70d-cb96-dd12-26b7-1380e3544bda:0:0: State change requested RUNNING --> FINISHED
2019-01-21 13:26:09,777 kk127.lab [23b9c70d-cb96-dd12-26b7-1380e3544bda:frag:0:0] DEBUG o.a.drill.exec.work.foreman.Foreman - 23b9c70d-cb96-dd12-26b7-1380e3544bda: cleaning up.
2019-01-21 13:26:09,779 kk127.lab [23b9c70d-cb96-dd12-26b7-1380e3544bda:frag:0:0] INFO  query.logger - {"queryId":"23b9c70d-cb96-dd12-26b7-1380e3544bda","schema":"","queryText":"show databases","start":1548105969411,"finish":1548105969778,"outcome":"COMPLETED","username":"mapr","remoteAddress":"172.30.11.185:58039"}
```

ihuzenko commented:
Hello @kkhatua , sorry for the late answer. I wanted to start review from check that fix works fine, but suddenly bumped into an issue. I wasn't able to submit query with my current locale, because it affects number formatting. 

1. Tried to use default number: 
![drill-6960_locale_error](https://user-images.githubusercontent.com/4282630/51838059-ef6dac80-230d-11e9-9a8e-354a8361b971.png)

2. Tried to remove comma from default number: 
![drill-6960_locale_error2](https://user-images.githubusercontent.com/4282630/51838111-11ffc580-230e-11e9-9698-520da447a278.png)

3. Tried to use own auto limit 50 :
![drill-6960_default_number_taken_instead](https://user-images.githubusercontent.com/4282630/51838150-2a6fe000-230e-11e9-9335-717d49cad563.png)

After those failed attempts I executed the line ```console.log(window.navigator.userLanguage || window.navigator.language)``` in JS console and got locale ```en-US``` in my chromium-browser. 

@kkhatua, can you please fix the blocking issue in order to unlock further review ? Thank you in advance, Igor.

ihuzenko commented:
Hello @kkhatua , sorry for the late answer. I wanted to start review from check that fix works fine, but suddenly bumped into an issue. I wasn't able to submit query with my current locale, because it affects number formatting. 

   1. Tried to use default number: 
![drill-6960_1](https://user-images.githubusercontent.com/4282630/51839839-0f539f00-2313-11e9-9e22-689d213a4a91.png)

   2. Tried to remove comma from default number: 
![drill-6960_2](https://user-images.githubusercontent.com/4282630/51840010-8ee16e00-2313-11e9-9bb6-b6b26f52c1e3.png)

   3. Tried to use own auto limit 50 :
![drill-6960_3](https://user-images.githubusercontent.com/4282630/51840016-9143c800-2313-11e9-9ab3-3d1e35bd1b16.png)

   After those failed attempts I executed the line ```console.log(window.navigator.userLanguage || window.navigator.language)``` in JS console and got locale ```en-US``` in my chromium-browser. 

   @kkhatua, can you please fix the blocking issue in order to unlock further review ? Thank you in advance, Igor.

kkhatua commented:
Thanks @ihuzenko .
I think i see the issue here being that the internal value didn't update after the field was changed. I'll fix that. 
But apart from that, I would think that `1,000` should not be considered as valid, considering that in Europe that is a decimal separator unlike the dot in the United States.

kkhatua commented:
@ihuzenko I narrowed down the cause of the bug to two things:
1. JQuery was loading the default (cached?) value of the textbox, so I had to force reading it from the root document object. That should fix this.
2. The locale issue is because Freemarker, by default, injects a comma as a thousands' separator in integer fields. I've forced it to not do that anymore, so it should resolve this too.
Please go ahead and review it.
(cc: @vdiravka )

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java:
```diff
@@ -46,15 +46,21 @@
+  public QueryWrapper(@JsonProperty("query") String query, @JsonProperty("queryType") String queryType, @JsonProperty("autoLimit") String autoLimit) {
     this.query = query;
     this.queryType = queryType.toUpperCase();
+    this.autoLimitRowCount = autoLimit.matches("[0-9]+") ? new Integer(autoLimit) : null;
```
As suggested in previous comment : 
```suggestion
   this.autoLimitRowCount = autoLimit != null && autoLimit.matches("[0-9]+") ? new Integer(autoLimit) : null;
```

kkhatua replied:
Done in the latest set of commits. Since it is a primitive data type, null check was removed

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java:
```diff
@@ -46,15 +46,21 @@
+
+  @JsonCreator
+  public QueryWrapper(@JsonProperty("query") String query, @JsonProperty("queryType") String queryType) {
+    this(query, queryType, "");
```
to avoid redundant pattern check : 
```suggestion
    this(query, queryType, null);
```

kkhatua replied:
Done in the latest set of commits

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryResources.java:
```diff
@@ -83,10 +83,14 @@ public QueryResult submitQueryJSON(QueryWrapper query) throws Exception {
       final String trimmedQueryString = CharMatcher.is(';').trimTrailingFrom(query.trim());
-      final QueryResult result = submitQueryJSON(new QueryWrapper(trimmedQueryString, queryType));
+      //Extract option of AutoLimit
+      String autoLimitValue = (autoLimit != null && autoLimit.matches("[0-9]+")) ? autoLimit : "";
```
To avoid double pattern check (here and in ```QueryWrapper```'s constructor), simply pass ```autoLimitValue``` to constructor of ```QueryWrapper``` and then within constructor set : 
```java 
this.autoLimitRowCount = autoLimit != null && autoLimit.matches("[0-9]+") ? new Integer(autoLimit) : null;
```

kkhatua replied:
Done in the latest set of commits

ihuzenko commented:
Hello @kkhatua . Thanks for fixing issues, I've checked changes and added review comments. Also I'd like to suggest remember user's choice and don't reset autoLimit checkbox and value after query submission. I think you could try [Window.sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage) for this.  Also please squash commits into one with name ```DRILL-6960: Auto Limit Wrapping should not apply to non-select query``` and description of changes below, like [here](https://github.com/apache/drill/pull/1631/commits/48f60f9a923222b4f8e42808085844f8fbf0be64)

kkhatua commented on exec/java-exec/src/main/resources/rest/static/js/querySubmission.js:
```diff
@@ -53,7 +53,18 @@ function doSubmitQueryWithUserName() {
+    } else {
+      let autoLimitValue=document.getElementById("autoLimit").value;
+      if (isNaN(autoLimitValue)) {
+        alert(autoLimitValue+ " is not a number. Please fill in a valid number");
```
We're already using the alert mechanism for empty username (when impersonation is enabled). In both cases, we want to indicate that the user cannot progress because of a missing username or invalid number. An error message on the query page might get missed.

ihuzenko replied:
ok, let's left it as is.

kkhatua replied:
This has been done as part of DRILL-7036 (#1644 )

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java:
```diff
@@ -69,14 +75,24 @@ public QueryType getType() {
+    return autoLimitRowCount;
+  }
+
+  public boolean isAutoLimitEnabled() {
```
Used locally, but only in one place. Will in-line the function instead.

kkhatua replied:
Don't need this any more, hence, removed in the latest set of commits

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +183,24 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
 
     return handler.getPlan(sqlNode);
   }
+
```
I am not familiar with how Calcite parsing works, but I would have that that wrapping the entire query, including `WITH` select statements, should work too. I'll test this and incorporate if nothing breaks. Thanks.

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +184,23 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
+  //Wrap with Auto Limit
+  private static SqlNode wrapWithAutoLimit(SqlConverter parser, QueryContext context, String sql) {
+    SqlNode sqlNode = parser.parse(sql);
+    if (context.isAutoLimitEnabled() && sqlNode.getKind().belongsTo(SqlKind.QUERY)) {
```
The ```context.isAutoLimitEnabled()``` is now checked before method invocation and may be removed here:
```suggestion
    if (sqlNode.getKind().belongsTo(SqlKind.QUERY)) {
```

kkhatua replied:
Done

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +184,23 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
+  private static SqlNode wrapWithAutoLimit(SqlConverter parser, QueryContext context, String sql) {
+    SqlNode sqlNode = parser.parse(sql);
+    if (context.isAutoLimitEnabled() && sqlNode.getKind().belongsTo(SqlKind.QUERY)) {
+      int autoLimitRowCount = context.getAutoLimitRowCount();
```
The StringBuilder code within ```if(){ ... }``` may be simplified to : 
```java
      String wrappedSql = String.format("-- [autoLimit: %2$d rows]%nSELECT * FROM (%n %1$s %n) LIMIT %2$d",
          sql, context.getAutoLimitRowCount());
      logger.debug("Query text to execute: {}", wrappedSql);
      sqlNode = parser.parse(wrappedSql);
```

kkhatua replied:
+1

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryResources.java:
```diff
@@ -83,10 +83,12 @@ public QueryResult submitQueryJSON(QueryWrapper query) throws Exception {
     try {
       final String trimmedQueryString = CharMatcher.is(';').trimTrailingFrom(query.trim());
-      final QueryResult result = submitQueryJSON(new QueryWrapper(trimmedQueryString, queryType));
+      final QueryResult result = submitQueryJSON(new QueryWrapper(trimmedQueryString, queryType, autoLimit)); //QueryWrapper will validate autoLimit
```
I think the comment is redundant and may be removed: 
```suggestion
      final QueryResult result = submitQueryJSON(new QueryWrapper(trimmedQueryString, queryType, autoLimit)); 
```

kkhatua replied:
Ok

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -47,6 +47,7 @@
 public class DrillSqlWorker {
   private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(DrillSqlWorker.class);
   private static final ControlsInjector injector = ControlsInjectorFactory.getInjector(DrillSqlWorker.class);
+  private static final String NEW_LINE = System.getProperty("line.separator");
```
May be removed after changing ```StringBuilder``` to ```String.format``` in ```wrapWithAutoLimit(parser, context, sql)``` method. See comments below for details.

kkhatua replied:
Done

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -273,6 +279,29 @@ public RemoteFunctionRegistry getRemoteFunctionRegistry() {
 
+  /**
+   * Check if auto-limiting of resultset is enabled
+   * @return
```
Done

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -273,6 +279,29 @@ public RemoteFunctionRegistry getRemoteFunctionRegistry() {
+
+  /**
+   * Returns the maximum size of auto-limited resultset
+   * @return
```
Done.

kkhatua commented:
@ihuzenko  
I've updated with changes and squashed and rebased onto current master. Please mark the Apache JIRA with a `ready-to-commit` tag if this looks alright.

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/user/UserWorker.java:
```diff
@@ -73,9 +73,13 @@ private static QueryId queryIdGenerator() {
+    return submitWork(connection, query, null);
+  }
+
+  public QueryId submitWork(UserClientConnection connection, RunQuery query, Integer autoLimitRowCount) {
```
For the case when `RunQuery` will contain `autoLimitRowCount`, there will be no need for adding a new method.

kkhatua replied:
Agreed.. let me take a look at modifying RunQuery

kkhatua replied:
Done in the latest set of commits

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +183,20 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
+
+  //Wrap with Auto Limit
+  private static SqlNode wrapWithAutoLimit(SqlConverter parser, QueryContext context, String sql) {
+    SqlNode sqlNode = parser.parse(sql);
```
Please replace this code with the following to avoid parsing the query several times.
```suggestion
    SqlNode sqlNode = parser.parse(sql);

    int fetch = context.getAutoLimitRowCount();

    if (sqlNode.getKind().belongsTo(SqlKind.QUERY)) {
      if (sqlNode.getKind() == SqlKind.ORDER_BY) {
        // merge SqlOrderBy
        SqlOrderBy orderBy = (SqlOrderBy) sqlNode;
        if (orderBy.fetch != null) {
          fetch = Math.min(Integer.valueOf(orderBy.fetch.toString()), fetch);
        }
        sqlNode = new SqlOrderBy(orderBy.getParserPosition(), orderBy.query, orderBy.orderList, orderBy.offset,
            SqlLiteral.createExactNumeric(Integer.toString(fetch), SqlParserPos.ZERO));
      } else {
        sqlNode = new SqlOrderBy(SqlParserPos.ZERO, sqlNode, SqlNodeList.EMPTY, null,
            SqlLiteral.createExactNumeric(Integer.toString(fetch), SqlParserPos.ZERO));
      }
    }
```

kkhatua replied:
This helps a lot! I've no knowledge of the Calcite SqlNodes, so this snippet helps.

vvysotskyi commented:
Regarding this fix and fix for DRILL-6050. I think allowing web-client to rewrite the query is not the best approach. A user may want to submit the query to see its profile for example, but with this feature, the profile will not correspond to the original query. Besides that, this fix does a lot of changes for web client on the server side.

It would be better to address this issue by limiting rows which should be obtained from the result set on the client side (see [WebUserConnection.sendData()](https://github.com/apache/drill/blob/master/exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebUserConnection.java#L138) method).

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -273,6 +279,29 @@ public RemoteFunctionRegistry getRemoteFunctionRegistry() {
+   * Returns the maximum size of auto-limited resultset
+   * @return Maximum size of auto-limited resultSet
+   */
+  public Integer getAutoLimitRowCount() {
```
The feature is specifically for WebUI submitted queries. A query option doesn't make sense because WebUI queries are "session-less" and the only reason we want to wrap with a limit is to prevent the Drillbit from being overloaded with holding an entire result set that does not get completely consumed by the HTTP client.

ihuzenko replied:
@kkhatua , I think adding new query scope option isn't a bad idea, because now same ```autoLimitRowCount``` and related methods are spread across lots of places, and if it's possible to minimize the duplication let's give it a try. Also options can have good long descriptions where you can mention all details about how it's used and  why it was added.

kkhatua replied:
Done in the latest set of commits

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java:
```diff
@@ -136,15 +136,30 @@
+   * @param queryRequest the query to execute
+   * @param autoLimitRowCount the number of rows to limit in execution (used for WebRequests)
+   */
+  public Foreman(final WorkerBee bee, final DrillbitContext drillbitContext,
```
I was reluctant to make too many changes to the protobuf, so I avoided modifying RunQuery.

kkhatua replied:
Done in the latest set of commits

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/ProfileWrapper.java:
```diff
@@ -150,6 +150,14 @@ private long tallyMajorFragmentCost(List<MajorFragmentProfile> majorFragments) {
+  }
+
+  public int getAutoLimit() {
+    return profile.getAutoLimit();
```
My statement above... I don't think the query option is the right thing to do, since this is a WebUI interaction feature.

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -84,9 +85,14 @@
+    this(session, drillbitContext, queryId, null);
+  }
+
+  public QueryContext(UserSession session, DrillbitContext drillbitContext, QueryId queryId, Integer autoLimit) {
```
I'm effectively creating an additional constructor with a modified signature, so as to not change anything else.  We cant go the "query option"  solution, so let me see if I can do without it.

kkhatua commented:
Let me take a look at `WebUserConnection` . I want to make sure that we don't hold all the results indefinitely if we get a hint (the autoLimit) that the entire resultset will not be consumed. From the description of the class, it looks like it might be able to do that.

kkhatua commented:
@vvysotskyi (cc @ihuzenko )
I took your advise and enforced the limit in the WebUserConnection. That way, I don't modify the query and no changes to the protobufs either. Thanks for pointing me to the approach as it is much cleaner.
This way, we mimic exactly what a JDBC driver would do after fetching a fixed number of records before closing the connection.
Please take some time to review the PR.

kkhatua commented:
@vvysotskyi (cc @ihuzenko )
I was discussing with Sorabh the issue of whether I can fetch only an initial small part of a resultset for a query and terminate it without consuming the remaining rows, and he pointed out that cancellation is the only way to do this, which is not a very graceful way of executing the query to fetch a limited set of rows. My current problem is that a massive join query was able to create the small dataset ready for the WebServer when I tried you approach of not building the resultset beyond what the limit was. However, I am still forced to iterate through the remaining incoming batches till the query completes. That means, while I have the benefit of not overload the WebServer (original intent of DRILL-6050 and DRILL-6477), the Foreman is still forced to run through all the remaining batches only to discard them. 
As a result, the Calcite approach suggested by Igor seems to be the best approach, because it also ensures that the Foreman does not process records that will not be sent to the Web/REST client.
I'm in the process of doing some final testing with the last commit on the Calcite-based solution, which are mostly UX related. Should I submit the commit ID to this PR or open a new PR for the review?

kkhatua commented:
**DRILL-6960: AutoLimit the size of ResultSet for a WebUI (or REST) client**
1. Check if username (if impersonation is enabled) and limits are valid inputs
2. Switch `help` from onclick to a hovering tooltip
3. Protobuf Changes for QueryProfile to indicate if a query was run with a 'Limit wrap'
  a. UserBitShared.proto
  b. Java files (after # 3.a)
  c. C++ files (after # 3.a)
4. Server is able to detect the submitted autoLimit and apply the wrapper around the input SQL. For this, it evaluates SqlNode.getKind() and only wrapping with limit if it is a QUERY. If autoLimit is not applicable, remove the value to avoid incorrectly marking the profile during execution.
5. WebUI will render a pinned warning indicating that the result set size was limited. This should help explain two identical looking queries with different runtimes, because on of them ran with an implicit Limit Wrap
6. Queries can be resubmitted from their profile page, for which the optional limit is automatically enabled set to the value for the original profile. 
(That means a user does not have to remember to enable the limit wrap with the same value again, and can focus on the query instead)
7. The results page indicates if the number of records being rendered was auto-limited

kkhatua commented:
**Screenshot of Error Handling:**
![image](https://user-images.githubusercontent.com/4335237/52496526-6b76b880-2b88-11e9-8ae7-a28a4784d96c.png)


**Screenshot of results for the query**
```sql
-- Simple Join with existing Limit 22 [Expected: Limit 13 applied to SQL]
select r.r_regionkey as r_regionkey, r.r_name as r_name, n.n_nationkey as n_nationkey, n.n_name as n_name 
from dfs.par100.region r, dfs.par100.nation n 
where r.r_regionkey=n.n_regionkey 
limit 22
```
![image](https://user-images.githubusercontent.com/4335237/52496646-c6101480-2b88-11e9-8df7-fdc012ba9c4b.png)


**Screenshot of profile with Limit applied (overriding the existing larger limit 22)**
![image](https://user-images.githubusercontent.com/4335237/52496795-374fc780-2b89-11e9-8f20-702f590fe861.png)

kkhatua commented:
Tested for the following queries with Limit = 13

```sql
--ShowDatabases [Expected: NoLimit on SQL]
show databases

--SqlKind: WITH [Expected: Limit applied to SQL]
with X1 (r_regionkey, r_name, n_nationkey, n_name) as 
  (select r.r_regionkey as r_regionkey, r.r_name as r_name, n.n_nationkey as n_nationkey, n.n_name as n_name from dfs.par100.region r, dfs.par100.nation n where r.r_regionkey=n.n_regionkey)
 select X1.n_name as nation, X1.r_name as region, count(*) as tally 
  from X1, dfs.par100.customer c 
  where X1.n_nationkey = c.c_nationkey
  group by X1.n_name, X1.r_name
  order by tally

--SqlKind: ORDER_BY [Expected: Limit applied to SQL]
select r.r_name as region, n.n_name as nation
 , count(*) as tally 
  from dfs.par100.region r, dfs.par100.nation n 
  where r.r_regionkey=n.n_regionkey
  group by n.n_name, r.r_name
  order by tally

-- SqlKind: UNION outputting 30 rows [Expected: Limit applied to SQL]
select r_name as name, r_regionkey as code from dfs.par100.region
UNION
select n_name as name, n_nationkey as code from dfs.par100.nation 

-- Simple Join outputting 25 rows [Expected: Limit applied to SQL]
select r.r_regionkey as r_regionkey, r.r_name as r_name, n.n_nationkey as n_nationkey, n.n_name as n_name from dfs.par100.region r, dfs.par100.nation n where r.r_regionkey=n.n_regionkey

-- Huge Join outputting 600M rows [Expected: Limit applied to SQL and 4-node execution in <2 minutes]
select * from dfs.par100.lineitem l, dfs.par100.orders o where o.o_orderkey = l.l_orderkey;

-- Simple Join with existing Limit 22 [Expected: Limit 13 applied to SQL]
select r.r_regionkey as r_regionkey, r.r_name as r_name, n.n_nationkey as n_nationkey, n.n_name as n_name from dfs.par100.region r, dfs.par100.nation n where r.r_regionkey=n.n_regionkey limit 22

-- Simple Join with existing Limit 11 [Expected: NoLimit on SQL]
select r.r_regionkey as r_regionkey, r.r_name as r_name, n.n_nationkey as n_nationkey, n.n_name as n_name from dfs.par100.region r, dfs.par100.nation n where r.r_regionkey=n.n_regionkey limit 11

--CreateTable [Expected: NoLimit on SQL]
create temporary table dummy ( nation )  as 
select n_name from dfs.par100.nation
```

kkhatua commented:
@ihuzenko / @vvysotskyi 
Please review the squashed commits. 
It includes changes requested by you and the PR summary is mentioned in the comment: https://github.com/apache/drill/pull/1608#issuecomment-461889067
Screenshots and test queries are in the following comments

ihuzenko commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +188,34 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
+
+  //Wrap with Auto Limit
+  private static SqlNode checkAndApplyAutoLimit(SqlConverter parser, QueryContext context, String sql) {
+    SqlNode sqlNode = parser.parse(sql);
```
Hello @kkhatua , I agree with logic implemented within the ```checkAndApplyAutoLimit(SqlConverter parser, QueryContext context, String sql)``` method. But it would be nice to make it more readable. Please refactor this method, as an example you can use my snippet below:
 
```java
  private static SqlNode checkAndApplyAutoLimit(SqlConverter parser, QueryContext context, String sql) {
    SqlNode sqlNode = parser.parse(sql);
    if (isAutoLimitShouldBeApplied(context, sqlNode)) {
      sqlNode = wrapWithAutoLimit(sqlNode, context);
    } else {
      context.disableAutoLimit();
    }
    return sqlNode;
  }

  private static boolean isAutoLimitShouldBeApplied(QueryContext context, SqlNode sqlNode) {
    return context.isAutoLimitEnabled() && sqlNode.getKind().belongsTo(SqlKind.QUERY)
        && (sqlNode.getKind() != SqlKind.ORDER_BY || isAutoLimitLessThanOrderByFetch((SqlOrderBy) sqlNode, context));
  }

  private static boolean isAutoLimitLessThanOrderByFetch(SqlOrderBy orderBy, QueryContext context) {
    return orderBy.fetch == null || Integer.parseInt(orderBy.fetch.toString()) > context.getAutoLimitRowCount();
  }

  private static SqlNode wrapWithAutoLimit(SqlNode sqlNode, QueryContext context) {
    SqlNumericLiteral autoLimitLiteral = SqlLiteral.createExactNumeric(context.getAutoLimitRowCount().toString(), SqlParserPos.ZERO);
    if (sqlNode.getKind() == SqlKind.ORDER_BY) {
      SqlOrderBy orderBy = (SqlOrderBy) sqlNode;
      return new SqlOrderBy(orderBy.getParserPosition(), orderBy.query, orderBy.orderList, orderBy.offset, autoLimitLiteral);
    }
    return new SqlOrderBy(SqlParserPos.ZERO, sqlNode, SqlNodeList.EMPTY, null, autoLimitLiteral);
  }
   
```
Note that I removed log.debug statements, I consider this one ```logger.debug("AutoLimit was{}applied", appliedLimit ? " " : " not ");``` redundant because you're logging something like this when filling QueryProfile. And this statement  
```logger.debug("Query text to execute: {}", sqlNode.toSqlString(HiveSqlDialect.DEFAULT));``` I assume is harmful , because it's serialization of sqlNode tree back to SQL which will be done for each query, even when debug log level is disabled.

kkhatua replied:
Good point about the serialization of the sqlNode. And since you've broken up the logic across multiple methods, we can do without the debug message. This helps. Thanks, @ihuzenko !

kkhatua commented:
@ihuzenko  (cc: @vvysotskyi ) Done the changes. Travis unit tests are passing on the private branch. I just fast-forwarded this PR's branch to that commit (https://travis-ci.org/kkhatua/drill/builds/492846830)
Functional and Advanced regressions tests have passed as well.

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java:
```diff
@@ -143,8 +144,9 @@ public Foreman(final WorkerBee bee, final DrillbitContext drillbitContext,
     closeFuture.addListener(closeListener);
-
-    this.queryContext = new QueryContext(connection.getSession(), drillbitContext, queryId);
+    //Extract any auto-limit for the resultSet in case of a WebUser invoked query
```
+1 Spoke with @vvysotskyi . Will switch to using RunQuery for retrieving the rowcount

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java:
```diff
@@ -57,6 +57,7 @@
 import org.apache.drill.exec.server.DrillbitContext;
 import org.apache.drill.exec.server.FailureUtils;
 import org.apache.drill.exec.server.options.OptionManager;
+import org.apache.drill.exec.server.rest.WebUserConnection;
```
Done.

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebUserConnection.java:
```diff
@@ -196,4 +198,28 @@ public void cleanupSession() {
+   * Indicates if resultset is autolimited
+   * @return
+   */
+  boolean isResultSetAutoLimited() {
```
Yes. It was part of an older design. Removed it now.

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java:
```diff
@@ -143,8 +144,9 @@ public Foreman(final WorkerBee bee, final DrillbitContext drillbitContext,
-
-    this.queryContext = new QueryContext(connection.getSession(), drillbitContext, queryId);
+    //Extract any auto-limit for the resultSet in case of a WebUser invoked query
+    Integer autoLimitRowCount = (connection instanceof WebUserConnection) ? ((WebUserConnection) connection).getAutoLimitRowCount() : null;
```
Yes. Will be using the `RunQuery queryRequest` now based on my conversation with Vova

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +188,36 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
   }
+
+  //Check if AutoLimit should be applied
+  private static boolean isAutoLimitShouldBeApplied(QueryContext context, SqlNode sqlNode) {
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +188,36 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
+    return sqlNode;
+  }
+
+  //Check if autoLimit is less than existing limit in the query
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryWrapper.java:
```diff
@@ -75,6 +81,10 @@ public QueryResult run(final WorkManager workManager, final WebUserConnection we
         .setResultsMode(QueryResultsMode.STREAM_FULL)
         .build();
 
+    if (autoLimitRowCount != null) {
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -84,9 +85,14 @@
+  public QueryContext(final UserSession session, final DrillbitContext drillbitContext, final QueryId queryId, final Integer autoLimit) {
     this.drillbitContext = drillbitContext;
     this.session = session;
+    this.autoLimitRowCount = autoLimit;
```
👍  will take care of this

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebUserConnection.java:
```diff
@@ -196,4 +198,28 @@ public void cleanupSession() {
+   * Sets an autolimit on the size of records to be sent back on the connection
+   * @param Max number of records to be sent back to WebServer
+   */
+  void autoLimitResultSet(Integer autoLimitSize) {
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebUserConnection.java:
```diff
@@ -196,4 +198,28 @@ public void cleanupSession() {
+
+  /**
+   * Sets an autolimit on the size of records to be sent back on the connection
+   * @param Max number of records to be sent back to WebServer
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +188,36 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
+        && (sqlNode.getKind() != SqlKind.ORDER_BY || isAutoLimitLessThanOrderByFetch((SqlOrderBy) sqlNode, context));
+  }
+
+  //Check and apply auto Limit
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +188,36 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
     return handler.getPlan(sqlNode);
   }
+
+  //Check if AutoLimit should be applied
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryResources.java:
```diff
@@ -177,6 +181,16 @@ public String getRowsPerPageValues() {
     }
+
+    //Used by results.ftl to indicate autoLimited resultset
+    public boolean resultsAutoLimited() {
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -183,4 +188,36 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
+    return orderBy.fetch == null || Integer.parseInt(orderBy.fetch.toString()) > context.getAutoLimitRowCount();
+  }
+
+  //Wrap the query with the autoLimit value
```
👍

kkhatua commented:
@ihuzenko (cc: @vvysotskyi ) are the changes sufficient?

vvysotskyi commented:
@kkhatua, could you please fix failing unit tests and travis job?

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/ProfileWrapper.java:
```diff
@@ -150,6 +152,18 @@ private long tallyMajorFragmentCost(List<MajorFragmentProfile> majorFragments) {
+    return profile.getAutoLimit();
+  }
+
+  public Integer getDefaultAutoLimit() {
```
👍

vvysotskyi replied:
And this

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/profile/ProfileWrapper.java:
```diff
@@ -64,10 +64,12 @@
   private final boolean onlyImpersonationEnabled;
   private Map<String, String> physicalOperatorMap;
   private final String noProgressWarningThreshold;
+  private final Integer defaultAutoLimit;
```
👍

vvysotskyi replied:
Looks like it wasn't addressed.

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebUserConnection.java:
```diff
@@ -196,4 +198,20 @@ public void cleanupSession() {
+   * Gets the max size of records to be sent back by the query
+   * @return Max number of records to be sent back to WebServer
+   */
+  public Integer getAutoLimitRowCount() {
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -273,6 +279,29 @@ public RemoteFunctionRegistry getRemoteFunctionRegistry() {
+   * Returns the maximum size of auto-limited resultset
+   * @return Maximum size of auto-limited resultSet
+   */
+  public Integer getAutoLimitRowCount() {
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -76,6 +76,7 @@
   private final SchemaTreeProvider schemaTreeProvider;
   /** Stores constants and their holders by type */
   private final Map<String, Map<MinorType, ValueHolder>> constantValueHolderCache;
+  private Integer autoLimitRowCount;
```
Done those changes and also for the WebUI validation

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -76,6 +76,7 @@
   private final SchemaTreeProvider schemaTreeProvider;
   /** Stores constants and their holders by type */
   private final Map<String, Map<MinorType, ValueHolder>> constantValueHolderCache;
+  private Integer autoLimitRowCount;
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -273,6 +279,29 @@ public RemoteFunctionRegistry getRemoteFunctionRegistry() {
+   * @return True if auto-limit is enabled
+   */
+  public boolean isAutoLimitEnabled() {
+    return autoLimitRowCount != null && autoLimitRowCount > 0;
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebUserConnection.java:
```diff
@@ -196,4 +198,20 @@ public void cleanupSession() {
+   * Sets an autolimit on the size of records to be sent back on the connection
+   * @param autoLimitRowCount Max number of records to be sent back to WebServer
+   */
+  void setAutoLimitRowCount(Integer autoLimitRowCount) {
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/WebUserConnection.java:
```diff
@@ -68,6 +68,8 @@
 
   public final List<String> metadata = new ArrayList<>();
 
+  private Integer autoLimitRowCount;
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -84,9 +85,14 @@
+    this(session, drillbitContext, queryId, null);
+  }
+
+  public QueryContext(final UserSession session, final DrillbitContext drillbitContext, final QueryId queryId, final Integer autoLimitRowCount) {
```
👍

kkhatua commented:
@vvysotskyi  the changes have been done, including the Regex fix for the rowcount value

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -85,8 +86,13 @@
+    this(session, drillbitContext, queryId, null);
+  }
+
+  public QueryContext(final UserSession session, final DrillbitContext drillbitContext, QueryId queryId, Integer autoLimit) {
```
Left the modifiers as they originally were

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java:
```diff
@@ -273,6 +279,21 @@ public RemoteFunctionRegistry getRemoteFunctionRegistry() {
   }
 
+  public boolean isAutoLimitEnabled() {
+    return (autoLimitRowCount != null);
```
👍

kkhatua commented on exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillSqlWorker.java:
```diff
@@ -135,7 +135,7 @@ private static PhysicalPlan getQueryPlan(QueryContext context, String sql, Point
     final SqlConverter parser = new SqlConverter(context);
     injector.injectChecked(context.getExecutionControls(), "sql-parsing", ForemanSetupException.class);
-    final SqlNode sqlNode = parser.parse(sql);
+    final SqlNode sqlNode = wrapWithAutoLimit(parser, context, sql);
```
Done in the latest set of commits

vvysotskyi commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/QueryResources.java:
```diff
@@ -177,6 +180,16 @@ public String getRowsPerPageValues() {
       return queryState;
     }
+
+    //Used by results.ftl to indicate autoLimited resultset
```
Could you please add a space before the comment: `//Used` -> `// Used` here and in other places?

kkhatua commented:
@vvysotskyi  for the case where the query's explicit limit matches the implicit auto-limit, the `QueryResources.TabularResult.isResultSetAutoLimited()` method has no qay of distinguishing if resultset was limited by the query explicitly or because of the implicit autolimit. This isn't really a bug, because semantically, the outcome is as expected. Implementing a "_fix_" would be non-trivial.

vvysotskyi commented:
I disagree that it is not a bug, limiting the result does not happen, so note shouldn't be displayed. Regarding the complexity of the fix, is it possible to obtain `autoLimit` value from query context there and check whether it is greater than zero?

kkhatua commented:
Yes, there is.. but I need to introduce a new method in the `UserConnection` object indicating that autoLimit was applied by the `DrillSqlWorker` (or not applied and disabled in the `QueryContext`).
I would need to set the value in the `UserConnection` here, just before I return the query Id, since the `DrillSqlWorker` has finished parsing (and disabled in the context): `org.apache.drill.exec.work.user.UserWorker.submitWork(UserClientConnection, RunQuery) QueryResource`  (https://github.com/apache/drill/blob/1a1cac0b2b9abd7dd90d614946f2101e5f3667f1/exec/java-exec/src/main/java/org/apache/drill/exec/work/user/UserWorker.java#L80)

kkhatua commented:
@vvysotskyi 
The latest commit (id: 12369c0) is able to achieve what you needed. We indicate on the initiating client if the autoLimit is set. The actual client response downstream (will read a value of zero if  `DrillSqlWorker` disabledAutolimit in the context,) wil detect that there was no autoLimit.

vvysotskyi commented:
@kkhatua, please rework the latest commit, it is not a good idea to add methods into `UserClientConnection` interface.

kkhatua commented:
@vvysotskyi , @ihuzenko 

I've opened #1714  as a new PR that includes fixes to this PR and also carries JDBC enhancements. Could you review that as an extension of this? I'll close this PR in the meanwhile, if that is OK.

kkhatua commented:
Closing this in favor of #1714 (DRILL-7048) which includes these fixes.
